src/catchup/IndexBucketsWork.cpp

```diff
@@ -0,0 +1,88 @@
+// Copyright 2022 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "IndexBucketsWork.h"
+#include "bucket/BucketIndex.h"
+#include "bucket/BucketList.h"
+#include "bucket/BucketManager.h"
+#include "util/HashOfHash.h"
+#include "util/types.h"
+#include "work/WorkWithCallback.h"
+#include <Tracy.hpp>
+
+namespace stellar
+{
+IndexBucketsWork::IndexBucketsWork(Application& app)
+    : Work(app, "index-bucketList", BasicWork::RETRY_NEVER)
+    , mBM(app.getBucketManager())
+{
+}
+
+BasicWork::State
+IndexBucketsWork::doWork()
+{
+    ZoneScoped;
+    if (!mChildrenSpawned)
+    {
+        auto indexBucket = [&](std::shared_ptr<Bucket> b) {
+            // Don't index empty bucket or buckets that are already being
+            // indexed. Sometimes one level's snap bucket may be another level's
+            // future bucket. The indexing job may have started but not
+            // finished, so check mIndexedBuckets. Some buckets may have already
+            // been indexed when assuming LCL state, so also check b->isIndexed.
+            if (b->getFilename().empty() || b->isIndexed() ||
+                mIndexedBuckets.find(b->getHash()) != mIndexedBuckets.end())
```

Comment:
is the last check for mIndexedBuckets needed? Isn't this already covered by b->isIndexed()?

Reply:
It is. There are a few weird edge cases where one level's snap bucket can be another level's future bucket. Before adding this check, we would kick off indexing jobs for all the buckets in the bucket list, including futures ready to resolve, and sometimes we would double index a bucket.